docs/code-organisation-guidelines.md

```diff
@@ -0,0 +1,65 @@
+# Code Organisation Guidelines
+
+The document specifies how code in the Cardano SL git repository should be organised. Consistent
+module naming schemes and file locations make the code relatively uniform so that it is easier for
+developers (internal and external), engineers working on formal specifications and auditors
+(internal and external) to navigate the code base.
+
+Currently, the code base does not conform to these guidelines but it is hoped that over time it
+can move from non-conforming to conforming. Making code conform to these guidelines in one big
+pull request or one long lived refactoring branch would be a huge amount of work and/or cause a
+lot of disruption to the regular flow of development. Instead, these guidelines should be applied
+as a series of small easy-to-review commits that can be merged relatively quickly.
+
+* The `cardano-sl` git repository includes a number of distinct Haskell packages like `core`, `util`,
+  `explorer`, `networking` and so on. All Haskell modules within each of these packages should be
+  given a unique name prefix. All modules within `core` would have a prefix `Pos.Core`, modules
+  within `util` would have a name prefix `Pos.Util`, modules within `networking` might have a prefix
+  of `Pos.Network` and so on. This means that when looking at the imports for a module, it is
+  immediately obvious which package each import comes from. As an example of a file that needs to be
+  renamed/moved the file `core/Pos/Aeson/Core.hs` should be renamed to `core/Pos/Core/Aeson.hs` so
+  that if it is imported in another module its name `Pos.Core.Aeson` would immediately suggest its
+  location in the source tree.
+
+* All library code for a Haskell packages should be in the `Pos` directory of that package.
```

Comment:
What does `Pos` mean? Why prefer it over, eg `Cardano` (other than the admirable goal of reducing import line length)?

If we enforce this requirement, then I'd want to have a common "utility" package where we could put stuff like `Control.Foobar.Extended` modules and other general purpose library-ish things that aren't quite worth spinning off into a Hackage library.

Reply:
`Pos` is Proof-of-stake which is kind of central to Cardano SL.

I think requiring `Pos` or `src/Pos` for extentions to existing concepts still makes sense. My naming the module `Pos.Util.Control.Foobar.Extended` you can immediately tell that is a Cardano SL specific extension to the `Foobar` concept and its located in the `util` package.

Reply:
That's good to know :smile: 

I'm convinced that the common prefix is a good idea, then.

Reply:
Let's go with `Cardano` then, I don't think there is any particular strong feeling in keeping `Pos` I believe 😉

Reply:
I think the only real reason to keep `Pos` would be to keep churn to a minimum, but if we're moving the code to the `src/` directory, the churn is already there. To help git track changes, we may want to move the code to `src/` in one commit and do the rename from `Pos` to `Cardano` in a second commit.